ssh/src/extension.ts

```diff
@@ -46,41 +53,226 @@ export function activate(context: vscode.ExtensionContext): void {
+			})
+			.on('error', (err) => {
+				reject(err);
+			})
+			.on('close', () => {
+				if (jump && jumpOptions) {
+					logger.log(`Closing ${getHostString(jumpOptions)} along with ${getHostString(destinationOptions)}`);
+					jump.end();
+				}
+			});
+		
+		jump
+			.on('ready', () => {
+				// Create an interactive shell with X11 forwarding and leave it open
+				// as long as VS Code is running so that we can point VS Code to its
+				// display.
+				createForwardingShell(jump, false);
+
+				jump.forwardOut('127.0.0.1', 12345, destinationOptions.host, destinationOptions.port, (err, stream) => {
+					if (err) {
+						reject(err);
+						return jump.end();
+					}
+					
+					logger.log(`Connecting to ${getHostString(destinationOptions)}`);
+					destination.connect({
+						sock: stream,
+						...getConnectConfig(destinationOptions)
+					});
+				});
+			})
+			.on('error', (err) => {
+				reject(err);
+			});
+
+		const preferConfig = getPreferConfig();
+		const jumpHost = getJumpHost();
+		let destinationOptions: ConnectOptions  = { 
+			username: options.username,
+			host: getServerHost() || options.host,
+			port: getServerPort() || options.port
+		}
 
-		logger.log(`Connecting to ${username}@${host} port ${port}`);
+		// Populate jump host falling back on server settings for user and port
+		let jumpOptions: ConnectOptions | null = null;
+		if (jumpHost) {
+			const jumpUser = getJumpUser();
+			const jumpPort = getJumpPort();
+			jumpOptions = {
+				username: jumpUser ? jumpUser : options.username,
+				host: jumpHost,
+				port: jumpPort ? jumpPort: options.port
+			}
+		}
 
-		conn.on('x11', handleX11);
+		if (preferConfig) {
+			const sshConfig = resolveHome(getSshConfig());
+			const host = getServerHost() || options.host;
+
+			logger.log(`Attempting to parse ${sshConfig} for ssh configuration of ${host}`);
+
+			try {
+				const data = fs.readFileSync(resolveHome(getSshConfig()))
+
+				const config = SSHConfig.parse(data.toString());
+
+				let hostConfig = null;
+				for (const line of config) {
+					if (line.param === 'Host') {
+						if (line.value === host) {
+							hostConfig = line.config;
+							break;
+						} else {
+							const hostNameMatch = line.config.find((line:any) => line.param === 'HostName' && line.value === host);
+							if (hostNameMatch) {
+								logger.log(`${host} matches ${line.value}`);
+								hostConfig = line.config;
+								break;
+							}
+						}
+					}
+				}
+
+				if (!hostConfig) {
+					throw new Error(`Could not find ${host}`);
+				}
+
+				const configUser = hostConfig.find((line:any) => line.param === 'User');
+				const configPort = hostConfig.find((line:any) => line.param === 'Port');
+				const configHost = hostConfig.find((line:any) => line.param === 'HostName');
+				const configProxy = hostConfig.find((line:any) => line.param === 'ProxyCommand');
+
+				destinationOptions = {
+					username: configUser ? configUser.value : options.username,
```

Comment:
Can simplify these using optional chaining and null coalescing.

```ts
{
  username: configUser?.value ?? options.username
}

Reply:
Mostly stuck in older versions of TS, haven't had a chance to leverage the chaining yet.